patches/server/0925-Paper-Plugins.patch

```diff
@@ -0,0 +1,1274 @@
++        this.register(file, file.getJarEntry(this.config), pluginPath);
++    }
++
++    protected abstract void register(JarFile file, JarEntry config, Path source) throws Exception;
++
++}
+diff --git a/src/main/java/io/papermc/paper/plugin/provider/type/paper/PaperPlugin.java b/src/main/java/io/papermc/paper/plugin/provider/type/paper/PaperPlugin.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..da931ff6057d3429e3e56c3df3d7abeb4a4e9d86
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/plugin/provider/type/paper/PaperPlugin.java
+@@ -0,0 +1,4 @@
++package io.papermc.paper.plugin.provider.type.paper;
++
++public interface PaperPlugin {
++}
+diff --git a/src/main/java/io/papermc/paper/plugin/provider/type/paper/PaperPluginProvider.java b/src/main/java/io/papermc/paper/plugin/provider/type/paper/PaperPluginProvider.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..460fdad978aedc2ebf5f785943b0f429524ae9be
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/plugin/provider/type/paper/PaperPluginProvider.java
+@@ -0,0 +1,251 @@
++package io.papermc.paper.plugin.provider.type.paper;
++
++import com.destroystokyo.paper.util.SneakyThrow;
++import io.papermc.paper.plugin.Initializer;
++import io.papermc.paper.plugin.provider.PluginProvider;
++import io.papermc.paper.plugin.provider.ProviderStatus;
++import io.papermc.paper.plugin.provider.ProviderStatusHolder;
++import io.papermc.paper.plugin.provider.configuration.PaperPluginConfiguration;
++import io.papermc.paper.plugin.provider.configuration.PluginConfiguration;
++import io.papermc.paper.plugin.provider.service.classloader.PaperPluginClassLoader;
++import io.papermc.paper.plugin.provider.service.classloader.PaperPluginClassLoaderStorage;
++import io.papermc.paper.plugin.provider.type.JavaPluginProvider;
++import org.bukkit.Bukkit;
++import org.bukkit.plugin.InvalidDescriptionException;
++import org.bukkit.plugin.InvalidPluginException;
++import org.bukkit.plugin.SimplePluginManager;
++import org.bukkit.plugin.java.JavaPlugin;
++import org.jetbrains.annotations.NotNull;
++import org.slf4j.Logger;
++import org.slf4j.LoggerFactory;
++import org.yaml.snakeyaml.error.YAMLException;
++
++import java.io.BufferedReader;
++import java.io.IOException;
++import java.io.InputStream;
++import java.io.InputStreamReader;
++import java.nio.file.Path;
++import java.util.jar.JarEntry;
++import java.util.jar.JarFile;
++
++public class PaperPluginProvider implements PluginProvider<PaperPlugin> {
++
++    public static PaperPluginProvider of(JarFile file, JarEntry config, Path pluginPath) throws Exception {
++        try (InputStream inputStream = file.getInputStream(config);
++             InputStreamReader inputStreamReader = new InputStreamReader(inputStream);
++             BufferedReader bufferedReader = new BufferedReader(inputStreamReader) // TODO: Is there a nicer way here?
++        ) {
```

Comment:
You don't need to close all theses streams manually just the lowest (the reader) and when java will close the stream via the try with stat it will also close all the children streams for you